redwoodjs/redwood
# feat: Mailer packages and studio tooling updates

This is probably as good an API I can make without getting people using it and telling me what needs improving. It will be much easier for people to trial if it's in canary or some minor RC. This is also a result of great input from other core team members!

Description:
This PR adds mailer packages - the core and a few example handlers and renderers. Studio is also updated to include a mail server for receiving mails during development and a live preview feature for rendering templates during development.

Outstanding:
1. ~~API side JSX support for JS projects~~
2. ~~Refactor to `send` and `sendWithoutRendering`~~
3. ~~Refactor the naming of everything~~
   More could be done here but let's get things in and get feedback on the general API first.
5. ~~Need to trial with something that doesn't render locally like SendGrid with their own dynamic templates~~
   This still needs done but I'm not sure about it yet. I think you'd probably have a separate handler for it and then call sendWithoutRendering on the mailer.
7. ~~Tests!~~
8. ~~Studio file watchers fire too many times~~
   Changed it a little and we can come back to it if it's a problem later.
10. ~~Inclusion strategy - what we include by default in new projects? Codemod bring old projects up to date with the default mailer content~~
   Let's decide this in a PR which updates the CRWA template and introduces the migration codemod.

Deep apologies for not doing a more incremental PR - pretty chunky PR when studio is included alongside the feature.

---

I'm not sure of the best way for this stage to be reviewed so extremely briefly; to use the new mailer:
1. Create a `api/src/lib/mailer.js/ts`
   This should `export const mailer = new Mailer({ ... })`. The config passed to the mailer involves a few properties and my recommendation would be to use typescript to explore it and play about with it.
2. Add your mail templates to `api/src/mail/**.*` as JSX or TSX files. These are just components which will you will pass to the mailer layer.
3. In your services or functions import the mailer like you would the db. The `send` and `sendWithoutRendering` methods are available on the mailer. Again my recommendation would be to explore the options to each via typescript. 

Using studio is done via `yarn rw exp studio`

An example project that uses the mailer:
https://github.com/Josh-Walker-GM/rw-test-mailer 

Here is also a notion doc that has some example instructions and that links to an informal video covering adding the mailer to the redwood blog test project:
https://verbose-alto-6ac.notion.site/Redwood-Mailer-canary-27dc0cc97bf04b358d5cf1738a81cf4c
